zio-http/jvm/src/main/scala/zio/http/netty/NettyBody.scala

```diff
@@ -40,31 +40,46 @@ object NettyBody extends BodyEncoding {
   private[zio] def fromAsync(
     unsafeAsync: UnsafeAsync => Unit,
     knownContentLength: Option[Long],
-    contentTypeHeader: Option[Header.ContentType] = None,
-  ): Body = AsyncBody(
-    unsafeAsync,
-    knownContentLength,
-    contentTypeHeader.map(_.mediaType),
-    contentTypeHeader.flatMap(_.boundary),
-  )
+    contentTypeHeader: () => Option[Header.ContentType] = () => None,
+  ): Body = {
+    lazy val cachedCt = contentTypeHeader()
+    AsyncBody(
+      unsafeAsync,
+      knownContentLength,
+      () => cachedCt.map(_.mediaType),
+      () => cachedCt.flatMap(_.boundary),
+    )
+  }
 
   /**
    * Helper to create Body from ByteBuf
    */
   def fromByteBuf(byteBuf: ByteBuf, contentTypeHeader: Option[Header.ContentType] = None): Body =
-    ByteBufBody(byteBuf, contentTypeHeader.map(_.mediaType), contentTypeHeader.flatMap(_.boundary))
+    fromByteBufLazy(byteBuf, () => contentTypeHeader)
+
+  private[zio] def fromByteBufLazy(
+    byteBuf: ByteBuf,
+    contentTypeHeader: () => Option[Header.ContentType] = () => None,
+  ): Body = {
+    lazy val cachedCt = contentTypeHeader()
+    ByteBufBody(byteBuf, () => cachedCt.map(_.mediaType), () => cachedCt.flatMap(_.boundary))
+  }
 
   override def fromCharSequence(charSequence: CharSequence, charset: Charset): Body =
     fromAsciiString(new AsciiString(charSequence, charset))
 
-  private[zio] final case class AsciiStringBody(
+  private[zio] final case class AsciiStringBody private[NettyBody] (
     asciiString: AsciiString,
-    override val mediaType: Option[MediaType] = None,
-    override val boundary: Option[Boundary] = None,
+    private val mediaType0: () => Option[MediaType] = () => None,
+    private val boundary0: () => Option[Boundary] = () => None,
```

Comment:
Boundary type is tricky because it comes in dynamically. However, it would be nice to eliminate the more complex API here and the `Function0` allocation. Did you measure to see if this has any impact?